notebook/notebook/notebook.py

```diff
@@ -131,10 +171,153 @@ def external_url_for(path):
+###### Methods that do not call the auth-gateway /verify method must not be directly public #######
+@app.route('/verify/<svc_name>/', methods=['GET'])
+def validate_get(svc_name):
+    access_token = request.args.get('authorization')
+    token = request.args.get('token')
+
+    if not access_token:
+        return forbidden()
+
+    resp = requests.get(f'http://{AUTH_GATEWAY}/verify',
+                        headers={'Authorization': f'Bearer {access_token}'})
+
+    if resp.status_code != 200:
+        return forbidden()
+
+    user_id = resp.headers.get('User')
+
+    if not user_id:
+        return forbidden()
+
+    try:
+        mysql_conn = pymysql.connect(
+        host=IP, port=PORT, user=USER, passwd=PASS, db=DB)
+
+        with mysql_conn.cursor() as cur:
+            cur.execute(f"SELECT svc_name FROM sessions WHERE user_id='{user_id}' AND token='{token}'")
+
+            rows = cur.fetchall()
+
+            if len(rows) != 1 or rows[0][0] != svc_name:
+                return forbidden()
+
+            return '', 200
+    except:
+        log.exception("/validate : GET : failed")
+        return '', 500
+
+
+@app.route('/api', methods=['GET'])
+def new_api_get():
+    user_id = request.headers.get('User')
+
+    if not user_id:
+        return forbidden()
+
+    try:
+        mysql_conn = pymysql.connect(
+        host=IP, port=PORT, user=USER, passwd=PASS, db=DB)
+
+        sql = f"SELECT name, svc_name, token FROM {TABLE} WHERE user_id='{user_id}' AND state<>'{DELETED_STATE}'"
+
+        rows = []
+        with mysql_conn.cursor() as cur:
+            cur.execute(sql)
+            rows = cur.fetchall()
+
+        return jsonify(rows), 200
+    except Exception as e:
+        log.exception(e)
+        return '', 500
+
+
+@app.route('/api/<token>', methods=['DELETE'])
+def delete_notebook(token):
+    user_id = request.headers.get("User")
+
+    # TODO: Is it possible to have a falsy token value and get here?
+    if not user_id or not token:
+        forbidden()
+
+    sqlDel = f"UPDATE {TABLE} SET state='{DELETED_STATE}' WHERE user_id='{user_id}' AND token='{token}'"
+    sqlFetch = f"SELECT svc_name, pod_name FROM {TABLE} WHERE user_id='{user_id}' AND token='{token}'"
+
+    try:
+        mysql_conn = pymysql.connect(
+            host=IP, port=PORT, user=USER, passwd=PASS, db=DB)
+
+        with mysql_conn.cursor() as cur:
+            rows_affected = cur.execute(sqlDel)
+
+            if rows_affected != 1:
+                log.exception(
+                    f'/api/<token> : DELETE : user_id {user_id} has multiple entries for token {token}')
+
+            cur.execute(sqlFetch)
+
+            svc_name, pod_name = cur.fetchone()
+
+            print("svc_name, pod_name", svc_name, pod_name)
+            delete_worker_pod(pod_name, svc_name)
+
+        mysql_conn.commit()
+        # If deletion operation fails because record doesn't exist,
+        # don't tell the user; they could exploit this
```

Comment:
no comments

Reply:
I disagree with this. If the comment isn't useful, tell me why.

Reply:
AFAICT, `Connection.commit` doesn't return anything so it's not clear how this text related to `commit`.

If there's an error we are purposely ignoring, we should `try .. catch ... return '', 200`. That makes the intention explicit in code.

Reply:
I see, then I should make the comment clearer. We return 200 no matter whether the deletion operation succeeded or not (but log an error that should in the future be pushed to a notification service / dashboard to notify us). The specific return value isn't really the point, happy to change that. I want to indicate why we don't give the user fine-grained reasons for failing operations.

In this case, if the user gave us the wrong token, we just tell them things are fine. It won't work, but we should provide an explicit / low information content error surface. The more detailed our responses are, the easier they are to exploit (by definition, they contain more information).